tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -144,13 +165,14 @@ public void CloseSpan(Span span)
             }
         }
 
-        public void SetSamplingPriority(int? samplingPriority, bool notifyDistributedTracer = true)
+        public void SetSamplingPriority(int? priority, int? mechanism = null, bool notifyDistributedTracer = true)
         {
-            _samplingPriority = samplingPriority;
+            _samplingPriority = priority;
+            _samplingMechanism = mechanism;
```

Comment:
I feel like we should call `AddSamplingMechanismTag` here with a parameter to set `replaceIfExists: true`. Otherwise, I don't know how the tag will be propagated via headers.

Reply:
On a separate note, what is the desired behavior when this is a downstream service and the user makes a sampling decision, but the upstream service has already propagated a sampling mechanism? Who wins?

Reply:
I found the following pseudocode in an RFC, is this up-to-date?

```
def update_sampling_priority(sampling_priority):
  if sampling_priority > 0:
    if trace.tags[_dd.p.dm] is not set:
        trace.tags[_dd.p.dm] = `-` + mechanism
  else:
    delete(trace.tags[_dd.p.dm])
```

Reply:
I refactored this to follow the RFC's pseudo-code more closely. See `SetSamplingMechanismTag()` and `RemoveSamplingMechanismTag()`.

It's a bit tricky because we do want to override the tag if we were the ones to set it (no upsteam service). Does this look ok to you?

Reply:
I did some more refactoring and merged `SetSamplingMechanismTag()` and `RemoveSamplingMechanismTag()` into `SetSamplingPriority()` and use that internally in `TraceContext`. Hopefully this is cleaner and easier to understand since we use the same method anywhere we are setting the sampling priority and mechanism.